### PR TITLE
DR2-1232 Add identifiers from Dynamo

### PR DESCRIPTION
**Pull request by MancunianSam**
This uses two new methods from https://github.com/nationalarchives/dr2-preservica-client/pull/72

The tests won't pass until this is deployed but this is ready for
review.

### CODE REVIEW

File: src/test/scala/uk/gov/nationalarchives/LambdaTest.scala [hunk 1]
```diff
@@ -229,6 +232,45 @@ class LambdaTest extends ExternalServicesTestUtils with MockitoSugar {
...
+
+            val mockLambda =
+              MockLambda(
+                convertFolderIdsAndRowsToListOfIoRows(rowsWithIdentifiers),
+                getIdentifiersForEntityReturnValues =
+                  IO(identifierFromPreservica.map(idp => IdentifierResponse("id", idp.identifierName, idp.value)))
+              )
+            mockLambda.handleRequest(mockInputStream, mockOutputStream, mockContext)
+
+            val updateCount = if (updateIdentifierRequest.isEmpty) 0 else 1
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/LambdaTest.scala:252]: 1. Couldn't you do `updateIdentifierRequest.length` instead?
2. Does this need to be a variable?

File: src/test/scala/uk/gov/nationalarchives/LambdaTest.scala [hunk 2]
```diff
@@ -229,6 +232,45 @@ class LambdaTest extends ExternalServicesTestUtils with MockitoSugar {
       }
   }
 
+  forAll(identifierScenarios) {
+    (identifierFromDynamo, identifierFromPreservica, addIdentifierRequest, updateIdentifierRequest) =>
+      {
+        "handleRequest" should s"add ${addIdentifierRequest.map(i => s"${i.identifierName}=${i.value}").mkString(" ")} " +
+          s"and update ${updateIdentifierRequest.map(i => s"${i.identifierName}=${i.value}").mkString(" ")}" in {
+            val rowsWithIdentifiers = ListMap.from(folderIdsAndRows.map { case (id, dynamoResponse) =>
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/LambdaTest.scala:240]: Could you not just do

```
folderIdsAndRows.take(1).map { case (id, dynamoResponse) => id -> dynamoResponse.copy(identifiers = identifierFromDynamo) }
```

**Review by techncl — COMMENTED** [review]
Looks good! Just some comments.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 3]
```diff
@@ -99,19 +148,48 @@ class Lambda extends RequestStreamHandler {
...
+      entity: Entity,
+      updatedIdentifier: Seq[UpdatedIdentifier],
+      addedIdentifiers: Seq[Identifier]
+  ): IO[Option[String]] = IO {
+    if (updatedIdentifier.isEmpty && addedIdentifiers.isEmpty) {
+      None
+    } else
+      Option {
+        val firstLine = generateSlackMessageFirstLine(preservicaUrl, entity)
+        firstLine + updatedIdentifier.headOption
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:163]: Yeah, makes sense. I've split it out into variables and added them at the end.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 4]
```diff
@@ -35,10 +37,33 @@ import upickle.default
 import java.io.{InputStream, OutputStream}
 import java.util.UUID
 import scala.io.Source
+import scala.jdk.CollectionConverters._
 
 class Lambda extends RequestStreamHandler {
   lazy val eventBridgeClient: DAEventBridgeClient[IO] = DAEventBridgeClient[IO]()
 
+  implicit val format: DynamoFormat[GetItemsResponse] = new DynamoFormat[GetItemsResponse] {
+    override def read(av: DynamoValue): Either[DynamoReadError, GetItemsResponse] = {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:46]: This is going now. It'll be replaced by https://github.com/nationalarchives/dr2-dynamo-formatters/pull/1

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 5]
```diff
@@ -35,10 +37,33 @@ import upickle.default
...
+      val map = av.toAttributeValue.m().asScala
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:47]: As above, I'll get rid of this.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 6]
```diff
@@ -35,10 +37,33 @@ import upickle.default
...
+      def value(name: String) = map.get(name).map(_.s())
+      def valueOrLeft(name: String) = value(name).toRight(MissingProperty)
+      for {
+        id <- valueOrLeft("id")
+        name <- valueOrLeft("name")
+      } yield {
+        val identifiers = map
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:54]: This is much neater. I've changed it in the dr2-dynamo-formatters.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 7]
```diff
@@ -35,10 +37,33 @@ import upickle.default
...
 class Lambda extends RequestStreamHandler {
   lazy val eventBridgeClient: DAEventBridgeClient[IO] = DAEventBridgeClient[IO]()
 
+  implicit val format: DynamoFormat[GetItemsResponse] = new DynamoFormat[GetItemsResponse] {
+    override def read(av: DynamoValue): Either[DynamoReadError, GetItemsResponse] = {
+      val map = av.toAttributeValue.m().asScala
+      def value(name: String) = map.get(name).map(_.s())
+      def valueOrLeft(name: String) = value(name).toRight(MissingProperty)
+      for {
+        id <- valueOrLeft("id")
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:51]: This has changed in the new code. It's called `id` in dynamo though.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 8]
```diff
@@ -35,10 +37,33 @@ import upickle.default
...
+        name <- valueOrLeft("name")
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:52]: Again, this has changed but it's `name` in Dynamo and it's not just the folder name

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 9]
```diff
@@ -377,4 +485,9 @@ object Lambda extends App {
       entity: Entity,
       updateEntityRequest: UpdateEntityRequest
   )
+
+  private[nationalarchives] case class UpdatedIdentifier(
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/Lambda.scala:490]: Fair point. I've gone with `IdentifierToUpdate`

File: src/test/scala/uk/gov/nationalarchives/LambdaTest.scala [hunk 10]
```diff
@@ -229,6 +232,45 @@ class LambdaTest extends ExternalServicesTestUtils with MockitoSugar {
       }
   }
 
+  forAll(identifierScenarios) {
+    (identifierFromDynamo, identifierFromPreservica, addIdentifierRequest, updateIdentifierRequest) =>
+      {
+        "handleRequest" should s"add ${addIdentifierRequest.map(i => s"${i.identifierName}=${i.value}").mkString(" ")} " +
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/LambdaTest.scala:238]: This is a good idea. I've done it but I've added the `if(addIdentfiersList.isEmpty()) "not add any...` bit into the function and then created two functions, one for add and one for update. It keeps the lines a bit shorter.
I've also changed the scalafmt conf to allow lines up to 150 so it stops sticking them everywhere.

File: src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala [hunk 11]
```diff
@@ -184,6 +189,29 @@ class ExternalServicesTestUtils
     )
   )
 
+  val identifierScenarios: TableFor4[List[Identifier], List[Identifier], List[Identifier], List[Identifier]] = Table(
+    ("identifierFromDynamo", "identifierFromPreservica", "addIdentifierRequest", "updateIdentifierRequest"),
+    (List(Identifier("Test1", "Value1")), List(Identifier("Test1", "Value1")), Nil, Nil),
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala:194]: I've saved them to variables. I had a bit of trouble with names that weren't so long they made things unreadable so I've mostly gone with `singleIdentifier1` `singleIdentifier2`

File: src/test/scala/uk/gov/nationalarchives/LambdaTest.scala (hunk 1 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/LambdaTest.scala:252]: Not sure why I did this, You're right, I've got rid of the variable and used `updateIdentifierRequest.length`

**Comment by MancunianSam** [comment]
The tests are failing because we need https://github.com/nationalarchives/dr2-dynamo-formatters/pull/1 and https://github.com/nationalarchives/dr2-preservica-client/pull/78 but I've made all of the suggested changes.

File: src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala [hunk 12]
```diff
@@ -184,16 +190,55 @@ class ExternalServicesTestUtils
     )
   )
 
+  private val singleIdentifier1: List[Identifier] = List(Identifier("Test1", "Value1"))
+  private val singleIdentifierDifferentValue: List[Identifier] = List(Identifier("Test1", "Value2"))
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala:194]: Maybe `singleIdentifier1DifferentValue`?

**Review by techncl — COMMENTED** [review]
Looks good, just a few minor comments.

File: src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala [hunk 13]
```diff
@@ -184,16 +190,55 @@ class ExternalServicesTestUtils
...
+  private val singleIdentifier2: List[Identifier] = List(Identifier("Test2", "Value2"))
+  private val multipleIdentifiers: List[Identifier] = singleIdentifier1 ++ singleIdentifier2
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala:196]: I've just read the comment below, ignore me.

File: src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala [hunk 14]
```diff
@@ -184,16 +190,55 @@ class ExternalServicesTestUtils
...
+      singleIdentifier1,
+      Nil,
+      singleIdentifierDifferentValue,
+      addIdentifiersDescription(Nil),
+      updateIdentifiersDescription(singleIdentifier1)
+    ),
+    (singleIdentifier1, singleIdentifier2, singleIdentifier1, Nil, addIdentifiersDescription(singleIdentifier1), updateIdentifiersDescription(Nil)),
+    (
+      multipleIdentifiers,
+      singleIdentifier3,
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala:213]: How does that help? It's testing an add and an update.

[the remark this answers is not in the document]
Comment (techncl) [reply src/test/scala/uk/gov/nationalarchives/testUtils/ExternalServicesTestUtils.scala:213]: Sorry, I missed this comment. Even though it's merged, I'll answer it. This and the other comment were a callback to the conversation we had about the order of the values; in the test, now, Preservica returns `Value3` and Dynamo updates it with `Value2` but it's no big deal as the test is doing what it's supposed to.

**Pull request merged**